### PR TITLE
Laboratory work #2, Polina Tokunova - 22FPL1

### PR DESCRIPTION
**Pull request by polsha24**

### CODE REVIEW

File: lab_2_tokenize_by_bpe/main.py [hunk 1]
```diff
@@ -14,6 +14,17 @@ def prepare_word(
     :param end_of_word: a token that signifies the end of word
     :return: preprocessed word
     """
+    if not isinstance(raw_word, str) \
+            or not isinstance(start_of_word, (str | None)) \
+            or not isinstance(end_of_word, (str | None)):
+        return None
+    if start_of_word is None:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:21]: too dificult, let's change the conditions

File: lab_2_tokenize_by_bpe/main.py [hunk 2]
```diff
@@ -36,6 +61,19 @@ def count_tokens_pairs(
...
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+    pair_freq = {}
+    for word, freq in word_frequencies.items():
+        for i in range(len(word) - 1):
+            token_pair = (word[i], word[i+1])
+            if token_pair not in pair_freq:
+                pair_freq[token_pair] = 0
+            if freq > 1:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:72]: very strange...

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 2 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 1 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py:20]: fixed

File: lab_2_tokenize_by_bpe/main.py [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
 
 def prepare_word(
-    raw_word: str, start_of_word: str | None, end_of_word: str | None
+        raw_word: str, start_of_word: str | None, end_of_word: str | None
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:8]: revert

File: lab_2_tokenize_by_bpe/main.py [hunk 4]
```diff
@@ -14,10 +14,22 @@ def prepare_word(
     :param end_of_word: a token that signifies the end of word
     :return: preprocessed word
     """
+    if not isinstance(raw_word, str) \
+            or not isinstance(start_of_word and end_of_word, (str | None)):
+        return None
+    if start_of_word is not None and end_of_word is not None:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:20]: oh no... Let's simplify this code

File: lab_2_tokenize_by_bpe/main.py [hunk 5]
```diff
@@ -14,10 +14,22 @@ def prepare_word(
...
+    if start_of_word is None and end_of_word is None:
+        return tuple([*list(raw_word)])
+    if start_of_word is None and end_of_word is not None:
+        return tuple([*list(raw_word), end_of_word])
+    if end_of_word is None and start_of_word is not None:
+        return tuple([start_of_word, *list(raw_word)])
+    return None
 
 
 def collect_frequencies(
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:31]: .

File: lab_2_tokenize_by_bpe/main.py [hunk 6]
```diff
@@ -26,53 +38,128 @@ def collect_frequencies(
...
+            token_pair = (word[i], word[i + 1])
+            if token_pair not in pair_freq:
+                pair_freq[token_pair] = 0
+            pair_freq[token_pair] += freq
+    return pair_freq
 
 
 def merge_tokens(
-    word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
+        word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:78]: .

File: lab_2_tokenize_by_bpe/main.py [hunk 7]
```diff
@@ -26,53 +38,128 @@ def collect_frequencies(
...
 ) -> dict[tuple[str, ...], int] | None:
     """
     Updates word frequency dictionary by replacing a pair of token with a merged one
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(pair, tuple):
+        return None
+    str_pair = pair[0] + pair[1]
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:88]: join

File: lab_2_tokenize_by_bpe/main.py [hunk 8]
```diff
@@ -26,53 +38,128 @@ def collect_frequencies(
...
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(num_merges, int):
+        return None
+    merges_count = 0
+    while merges_count < num_merges:
+        pair_freq = count_tokens_pairs(word_frequencies)
+        if pair_freq is None:
+            return None
+        if pair_freq != {}:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:118]: bad check

File: lab_2_tokenize_by_bpe/main.py [hunk 9]
```diff
@@ -26,53 +38,128 @@ def collect_frequencies(
...
+                merges_count += 1
+                break
+            continue
+        break
+    return word_frequencies
 
 
 def get_vocabulary(
-    word_frequencies: dict[tuple[str, ...], int], unknown_token: str
+        word_frequencies: dict[tuple[str, ...], int], unknown_token: str
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:134]: .

File: lab_2_tokenize_by_bpe/main.py [hunk 10]
```diff
@@ -26,53 +38,128 @@ def collect_frequencies(
...
+    sorted_tokens_set = sorted(tokens_set, key=lambda item: (-len(item), item))
+    vocabulary = {}
+    for i, value in enumerate(sorted_tokens_set):
+        vocabulary.update({value: i})
+    return vocabulary
 
 
 def decode(
-    encoded_text: list[int] | None, vocabulary: dict[str, int] | None, end_of_word_token: str | None
+        encoded_text: list[int] | None,
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:160]: .

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Please, revert code that you do not have to change

File: lab_2_tokenize_by_bpe/main.py [hunk 11]
```diff
@@ -14,6 +14,17 @@ def prepare_word(
     :param end_of_word: a token that signifies the end of word
     :return: preprocessed word
     """
+    if isinstance(raw_word, str) and isinstance(start_of_word and end_of_word, (str | None)):
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:17]: 1. erly return
2. too much code... simplify

File: lab_2_tokenize_by_bpe/main.py [hunk 12]
```diff
@@ -58,6 +107,26 @@ def train(
...
+        if not len(pair_freq) == 0:
+            sorted_dict = dict(
+                sorted(pair_freq.items(),
+                       key=lambda item: (-item[1], -len("".join(item[0])), "".join(item[0]))))
+            for k in sorted_dict.keys():
+                word_frequencies = merge_tokens(word_frequencies, k)
+                if word_frequencies is None:
+                    return None
+                merges_count += 1
+                break
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:126]: too many brake/continue. Revert the condition

File: lab_2_tokenize_by_bpe/main.py [hunk 13]
```diff
@@ -69,6 +138,21 @@ def get_vocabulary(
...
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(unknown_token, str):
+        return None
+    tokens_set = set()
+    tokens_set.add(unknown_token)
+    for word in word_frequencies.keys():
+        for token in word:
+            tokens_set.add(token)
+            if len(token) > 1:
+                for symbol in token:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:149]: for in if in for in for... simplify

File: lab_2_tokenize_by_bpe/main.py [hunk 14]
```diff
@@ -81,10 +165,23 @@ def decode(
...
+            or not isinstance(vocabulary, dict) \
+            or not isinstance(end_of_word_token, (str | None)):
+        return None
+    decoded_list = []
+    for code in encoded_text:
+        for k, value in vocabulary.items():
+            if code == value:
+                decoded_list.append(k)
+    decoded_text = "".join(decoded_list)
+    final_text = decoded_text.replace('</s>', ' ')
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:178]: Do not use this token as constant

File: lab_2_tokenize_by_bpe/main.py [hunk 15]
```diff
@@ -105,11 +202,11 @@ def load_vocabulary(vocab_path: str) -> dict[str, int] | None:
 
 
 def encode(
-    original_text: str,
-    vocabulary: dict[str, int] | None,
-    start_of_word_token: str | None,
-    end_of_word_token: str | None,
-    unknown_token: str,
+        original_text: str,
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:205]: revert

File: lab_2_tokenize_by_bpe/main.py [hunk 16]
```diff
@@ -132,7 +229,7 @@ def collect_ngrams(text: str, order: int) -> list[tuple[str, ...]] | None:
 
 
 def calculate_precision(
-    actual: list[tuple[str, ...]], reference: list[tuple[str, ...]]
+        actual: list[tuple[str, ...]], reference: list[tuple[str, ...]]
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:232]: revert

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 10 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py]: fix

File: lab_2_tokenize_by_bpe/main.py (hunk 9 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 8 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 7 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 6 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 4 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 3 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 14 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py:178]: done

File: lab_2_tokenize_by_bpe/main.py (hunk 16 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py:232]: done

File: lab_2_tokenize_by_bpe/main.py (hunk 15 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py:205]: done

File: lab_2_tokenize_by_bpe/main.py (hunk 12 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py:126]: done

File: lab_2_tokenize_by_bpe/main.py (hunk 11 above)
Comment (polsha24) [reply lab_2_tokenize_by_bpe/main.py]: done

**Pull request closed without merging**